datafusion/expr/src/logical_plan/plan.rs

```diff
@@ -369,8 +369,18 @@ impl LogicalPlan {
                 // The join keys in using-join must be columns.
                 let columns =
                     on.iter().try_fold(HashSet::new(), |mut accumu, (l, r)| {
-                        accumu.insert(l.try_into_col()?);
-                        accumu.insert(r.try_into_col()?);
+                        let Some(l) = l.try_as_col().cloned() else {
+                            return internal_err!(
```

Comment:
I think making the error here (rather than in `try_into_col`) results in a more specific error message that will be easier to debug if we ever hit it